engine: 23.0.0 release notes (#16421)
docker/docs: engine/release-notes/index.md

- Added a `--quiet` option to `docker create` and `docker run` commands to suppress output when pulling an image. [docker/cli#3377](https://github.com/docker/cli/pull/3377)
- Added a `--force` option to `docker network rm` subcommand. Causes CLI to return a 0 exit code even if the network doesn't exist. Has no effect on the server-side procedure for removing a network. [docker/cli#3547](https://github.com/docker/cli/pull/3547)
- Added a `--signal` option to `docker stop` and `docker restart`. [docker/cli#3614](https://github.com/docker/cli/pull/3614)
- Added a `-v/--version` flag to `docker-proxy`. [moby/moby#44703](https://github.com/moby/moby/pull/44703)
- Plugins are now discoverable in well-known user-level paths when the daemon is running in rootless mode. [moby/moby#44778](https://github.com/moby/moby/pull/44778)
- The daemon now handles common alternate JSON encodings in the JSON configuration file gracefully, and reports useful errors. [moby/moby#44777](https://github.com/moby/moby/pull/44777), [moby/moby#44832](https://github.com/moby/moby/pull/44832)

>   - JSON configuration with a UTF-8 byte order mark is accepted.

Reviewer: Let's do `UTF-8 with a byte order mark is accepted.` -- I was rephrasing here and it got a bit redundant.

- Set default sandbox size for WCOW to 127GB, and ensure storage-opts is honoured on all paths under WCOW. [moby/moby#41636](https://github.com/moby/moby/pull/41636)
- Removed the plugin section from the containerd configuration file `(/var/run/docker/containerd/containerd.toml)`. [moby/moby#41675](https://github.com/moby/moby/pull/41675)
- Reject `null` manifests during tar import. [moby/moby#41842](https://github.com/moby/moby/pull/41842)
- Added shim config for custom runtimes for plugins. [moby/moby#41854](https://github.com/moby/moby/pull/41854)
- Container health checks now resume when the daemon is restarted. [moby/moby#41935](https://github.com/moby/moby/pull/41935)
- Quota is no longer disabled on cleanup of the `btrfs` driver. [moby/moby#42273](https://github.com/moby/moby/pull/42273)
- (Accessible) host devices can now be mounted in `--privileged` rootless containers. [moby/moby#42638](https://github.com/moby/moby/pull/42638)
- Fixed incorrect handling of `**/foo` pattern in `fileutils`. [moby/moby#42676](https://github.com/moby/moby/pull/42676)

> - Added the ability to pass a `--platform` flag on `docker import` when importing an archive [moby/moby#43103](https://github.com/moby/moby/pull/43103)

Reviewer: ```suggestion
- Extended `docker import --platform` to allow marking an imported image as a foreign architecture. [moby/moby#43103](https://github.com/moby/moby/pull/43103)
```

- Validation of CPU Realtime options is now performed when the daemon starts instead of performing validations for each individual container, so that we can fail early. [moby/moby#43131](https://github.com/moby/moby/pull/43131)
- Freeze the `namesgenerator` package against new additions. Users will have to be satisfied with the existing 25359 adjective-name combinations. [moby/moby#43210](https://github.com/moby/moby/pull/43210)

> - Attach `containers/{ID}/attach/ws` only to streams according by stdin, stdout and stderr parameters. [moby/moby#43322](https://github.com/moby/moby/pull/43322)

Reviewer: ```suggestion
- API: `/containers/{id}/attach/ws` only to streams according by stdin, stdout and stderr parameters on API >= 1.42. [moby/moby#43322](https://github.com/moby/moby/pull/43322)
```

- Fixed UDP traffic in containers not working after the container is restarted under sustained traffic. [moby/moby#43409](https://github.com/moby/moby/pull/43409)

> - Added support for pulling images with custom amd64 variants that are used by the latest Go, GCC, LLVM etc. [moby/moby#43434](https://github.com/moby/moby/pull/43434)

Reviewer: ```suggestion
- Added support for pulling images with amd64 microarchitecture feature levels as recently supported by Go, GCC, LLVM, and other compiler toolchains. [moby/moby#43434](https://github.com/moby/moby/pull/43434)
```

- Improved validation of invalid JSON requests in the API. [moby/moby#43463](https://github.com/moby/moby/pull/43463)

> - Mitigated the impact of slow `exec` starts on health checks. Health check timeout now applies only to the duration that the health-check command is running. The time it takes to start the command no longer counts against the timeout. [moby/moby#43480](https://github.com/moby/moby/pull/43480)

Reviewer: ```suggestion
- Mitigated the impact of slow `exec` starts on health checks. The health check timeout now only applies to the duration that the health-check command is running. The time it takes to start the command no longer counts against the timeout. [moby/moby#43480](https://github.com/moby/moby/pull/43480)
```

- Console `tty` size is set immediately on creation. [moby/moby#43593](https://github.com/moby/moby/pull/43593), [moby/moby#43622](https://github.com/moby/moby/pull/43622)

> - Fixed `daemon.Shutdown`, and `daemon.Cleanup` not cleaning up `overlay2` mounts. [moby/moby#43659](https://github.com/moby/moby/pull/43659)

Reviewer: ```suggestion
- Fixed `overlay2` mounts not being cleaned up after failed container starts or daemon shutdown. [moby/moby#43659](https://github.com/moby/moby/pull/43659)
```

- Match manifest list resolution with `containerd`. [moby/moby#43675](https://github.com/moby/moby/pull/43675)

> - Skip `firewalld` management for RootlessKit in libnetwork. [moby/moby#43813](https://github.com/moby/moby/pull/43813)

Reviewer: ```suggestion
- Skip use of `firewalld` for networking when the daemon is running in rootless mode. [moby/moby#43813](https://github.com/moby/moby/pull/43813)
```

> - Custom NAT networks are now re-created after daemon restart if missing from HNS on Windows. [moby/moby#43858](https://github.com/moby/moby/pull/43858)

Reviewer: ```suggestion
- Custom NAT networks are now re-created after daemon restart if missing on Windows. [moby/moby#43858](https://github.com/moby/moby/pull/43858)
```

- Fixed a failure to create the `DOCKER-USER` ip6tables chain. [moby/moby#44845](https://github.com/moby/moby/pull/44845)
- Fixed a failure to clean up iptables rules when the `ip6tables` command isn't available. [moby/moby#44727](https://github.com/moby/moby/pull/44727)
- Fixed an issue where some iptables NAT rules weren't cleaned up after enabling the userland proxy. [moby/moby#44811](https://github.com/moby/moby/pull/44811)
- Fixed a potentially leaked process in rare situations where cleaning up a failed attempt to start a container was mishandled. [moby/moby#44400](https://github.com/moby/moby/pull/44400)
- Fixed the `CreatedAt` time of a volume reflecting initialization and not creation. [moby/moby#44725](https://github.com/moby/moby/pull/44725)
- Fixed an issue where the CLI incorrectly reported an incompatible server instead of an unreachable server in some commands. [docker/cli#3901](https://github.com/docker/cli/pull/3901), [docker/cli#3904](https://github.com/docker/cli/pull/3904)

> - Fixed broken completion of volumes in ZSH. [docker/cli#2998](https://github.com/docker/cli/pull/2998)

Reviewer: ```suggestion
- Fixed broken completion of volumes in Zsh. [docker/cli#2998](https://github.com/docker/cli/pull/2998)
```

Apparently they capitalize it 'Zsh' in the Zsh project.

- API: Removed the deprecated `KernelMemory` option from `POST /containers/create` on API version >= 1.42. [moby/moby#43214](https://github.com/moby/moby/pull/43214)
  - [Deprecation notice](../deprecated.md#kernel-memory-limit)

### Deprecated

- Require Windows Server RS5 / LTSC 2019 (build 17763) as the minimum to run the daemon. [moby/moby#43254](https://github.com/moby/moby/pull/43254)
- Deprecated `BuilderSize` in API versions >= 1.42 [moby/moby#42608](https://github.com/moby/moby/pull/42608)
- Deprecated `BuildCache.Parent` in favor of the newly introduced `BuildCache.Parents` in API version >= 1.42. [moby/moby#43908](https://github.com/moby/moby/pull/43908)
- Deprecated `pkg/urlutil`, moving the implementation to `builder/remotecontext/urlutil`. [moby/moby#43477](https://github.com/moby/moby/pull/43477)

### Upgrades

- Upgrade Go to `1.19.5`. [docker/cli#3958](https://github.com/docker/cli/pull/3958), [moby/moby#44794](https://github.com/moby/moby/pull/44794)
- Upgrade `rootlesskit` to `v0.14.4`. [moby/moby#42708](https://github.com/moby/moby/pull/42708)

> - Upgrade `buildkit` to `v0.10.0`. [moby/moby#43239](https://github.com/moby/moby/pull/43239)

Reviewer 2: It's BuildKit 0.10.6 actually: https://github.com/moby/moby/blob/de3f216113f27708006923376826686db3783ed4/vendor.mod#LL52C27-L52C34

```suggestion
- Upgrade `buildkit` to `v0.10.6`. [moby/moby#43239](https://github.com/moby/moby/pull/43239)
```

Not sure if we should put all issues for each update. I guess the first one introducing BuildKit 0.10 is fine.

- Upgrade `swarmkit` to `v2.0.0-20230119195359-904c221ac281`. [moby/moby#44858](https://github.com/moby/moby/pull/44858)

> - Upgrade `containerd` to `v1.6.15`. [moby/moby#44766](https://github.com/moby/moby/pull/44766), [moby/moby#44769](https://github.com/moby/moby/pull/44769)

Reviewer 3: ```suggestion
- Upgrade `containerd` to `v1.6.16`. [moby/moby#44766](https://github.com/moby/moby/pull/44766), [moby/moby#44769](https://github.com/moby/moby/pull/44769), [moby/moby#44881](https://github.com/moby/moby/pull/44881)
```

Reviewer 3: Do we need to mention that we updated the bundled `buildx` version to v0.10.2 as well?

Author: Yes that would probably be good. Technically that is not an "upgrade", right, but a new addition that we could mention under the "Set Buildx as the default builder" bullet in the ## New section?

Reviewer: Buildx came with 20.10 so it's not new, it was just aliased to the default this release.

- Fixed an issue where some iptables NAT rules weren't cleaned up after enabling the userland proxy. [moby/moby#44811](https://github.com/moby/moby/pull/44811)
- Fixed a potentially leaked process in rare situations where cleaning up a failed attempt to start a container was mishandled. [moby/moby#44400](https://github.com/moby/moby/pull/44400)
- Fixed the `CreatedAt` time of a volume reflecting initialization and not creation. [moby/moby#44725](https://github.com/moby/moby/pull/44725)
- Fixed an issue where the CLI incorrectly reported an incompatible server instead of an unreachable server in some commands. [docker/cli#3901](https://github.com/docker/cli/pull/3901), [docker/cli#3904](https://github.com/docker/cli/pull/3904)
- Fixed broken completion of volumes in Zsh. [docker/cli#2998](https://github.com/docker/cli/pull/2998)
- Improved output of `docker context` when an invalid context is present. [docker/cli#3847](https://github.com/docker/cli/pull/3847)
- Removed ANSI decoration of CLI help annotations when the output isn't a TTY, and added a newline for readability. [docker/cli#3973](https://github.com/docker/cli/pull/3973)

> - Add `docker container remove` as an alias for `docker container rm`. [docker/cli#3986](https://github.com/docker/cli/pull/3986)

Reviewer: ```suggestion
- Added `docker container remove` as an alias for `docker container rm`. [docker/cli#3986](https://github.com/docker/cli/pull/3986)
```

Whoops, I forgot our chosen tense. LGTM after that little tweak.

Author: I forgot it too - actually we should prefer present tense so I've changed it throughout 🙈

  - /release-notes/docker-engine/
---

This page describes the latest changes, additions, known issues, and fixes for Docker Engine.

For more information about:

- Deprecated and removed features, see [Deprecated Engine Features](../deprecated.md).
- Changes to the Engine API, see [Engine API version history](../api/version-history.md).

## 23.0.0

{% include release-date.html date="2023-02-01" %}

For a full list of pull requests and changes in this release, refer to the relevant GitHub milestones:

- [docker/cli, 23.0.0 milestone](https://github.com/docker/cli/milestone/51?closed=1)
- [moby/moby, 23.0.0 milestone](https://github.com/moby/moby/milestone/91?closed=1)

### New

- Set Buildx and BuildKit as the default builder on Linux. [moby/moby#43992](https://github.com/moby/moby/pull/43992)

  - Alias `docker build` to `docker buildx build`. [docker/cli#3314](https://github.com/docker/cli/pull/3314)

>   - The legacy builder can still be used by explicitly setting `DOCKER_BUILDKIT=0`.

Reviewer: Maybe we should add a bullet with a link to https://docs.docker.com/build/building/multi-stage/#differences-between-legacy-builder-and-buildkit in light of some of the confusion on Slack?

Author: good idea! will add this for context

This document describes the latest changes, additions, known issues, and fixes
for Docker Engine.

> ## 23.0.0

Reviewer 3: Been thinking about this for some time, but never started working on that.

First; we should consider doing a _rename_ of `index.md` to `20.10.md` (and fixing the links to the page) in a separate commit. That commit should _not_ add a new `index.md`, only a rename. That way, `git` will properly consider it a "rename", and preserve git history on the `20.10.md` file.

Second (this is the thing I wanted to work on, but didn't get to): we should use a similar approach as we take for the API reference, and make `index.md` (or have a `latest.md`) _redirect_ to the versioned release note (so currently `/engine/release-notes/` would redirect to `/engine/release-notes/20.10/`). That way we have proper "permalinks" for release notes (which may be referenced in (e.g.) CVE databases etc), but continue to have an easy-to-find URL for "what's the latest release note?".

Currently, we effectively break _all_ links to release notes when we do a new release, which isn't "great" :joy:

Author: argh, yes I noticed too late that I'd messed up the git records for these changes. I will fix it 😅

Author: regarding the latest/index redirect, we could also have the `index` file semantically represent a list page of the release notes, and always have each release note file versioned in the filename. Then we don't have to mess with redirects at all (unless we specifically want a `latest` endpoint)

> {% include release-date.html date="2023-02-01" %}

Reviewer 3: Yes, we don't have a final date yet. Probably we can keep this PR in "draft" while we're reviewing.

### New

- Set Buildx and BuildKit as the default builder on Linux. [moby/moby#43992](https://github.com/moby/moby/pull/43992)
  - Alias `docker build` to `docker buildx build`. [docker/cli#3314](https://github.com/docker/cli/pull/3314)
  - The legacy builder can still be used by explicitly setting `DOCKER_BUILDKIT=0`.
- Added support for pulling `zstd` compressed layers. [moby/moby#41759](https://github.com/moby/moby/pull/41759), [moby/moby#42862](https://github.com/moby/moby/pull/42862)
- Added support for alternate OCI runtimes on Linux, compatible with the containerd runtime v2 API. [moby/moby#43887](https://github.com/moby/moby/pull/43887), [moby/moby#43993](https://github.com/moby/moby/pull/43993)

> - Added support for the v2 containerd `runhcs` shim on Windows (off by default). [moby/moby#42089](https://github.com/moby/moby/pull/42089)

Reviewer: ```suggestion
- Added support for the containerd `runhcs` shim on Windows (off by default). [moby/moby#42089](https://github.com/moby/moby/pull/42089)
```

- Added support for describing Windows devices using the syntax `IDType://ID`.[moby/moby#43368](https://github.com/moby/moby/pull/43368)
- Added `RootlessKit`, `slirp4netns`, and `VPNKit` version reporting. [moby/moby#42330](https://github.com/moby/moby/pull/42330)
- Added experimental support for SwarmKit cluster volumes (CSI). [moby/moby#41982](https://github.com/moby/moby/pull/41982)
  - CLI: Added cluster volume (CSI) options to `docker volume`. [docker/cli#3606](https://github.com/docker/cli/pull/3606)
  - CLI: Added cluster volume (CSI) support to `docker stack`. [docker/cli#3662](https://github.com/docker/cli/pull/3662)
- Added support for SwarmKit jobs in `docker stack deploy`. [docker/cli#2907](https://github.com/docker/cli/pull/2907)
- Added the `docker stack config` command to output the merged and interpolated config files as utilized by `stack deploy`. [docker/cli#3544](https://github.com/docker/cli/pull/3544)
- Added a new `docker context show` command that prints the name of the current context. [docker/cli#3567](https://github.com/docker/cli/pull/3567)

> - Added a `--format=json` option to `docker inspect` and `docker ls`. [docker/cli#2936](https://github.com/docker/cli/pull/2936)

Reviewer 3: Or we should make it more generic; if we did everything well, all `--format` flags (on any command) should now support `json` as special value (which is effectively a shorthand for `{{ json . }}`)

> - Added a `--format=json` option to `docker inspect` and `docker ls`. [docker/cli#2936](https://github.com/docker/cli/pull/2936)

Author: oops, that's my bad, thanks for spotting @tianon !

> - Added a `--format=json` option to `docker inspect` and `docker ls`. [docker/cli#2936](https://github.com/docker/cli/pull/2936)

Reviewer 4: Yeah, I like that @thaJeztah :smile:

- Added a `--force` option to `docker network rm` subcommand. Causes CLI to return a 0 exit code even if the network does not exist. Has no effect on the server-side procedure for removing a network. [docker/cli#3547](https://github.com/docker/cli/pull/3547)
- Added a `--signal` option to `docker stop` and `docker restart`. [docker/cli#3614](https://github.com/docker/cli/pull/3614)
- Allow use of `STOPSIGNAL` via `docker commit`. [moby/moby#43369](https://github.com/moby/moby/pull/43369)
- Added a new option to the `awslogs` log driver to allow skipping log stream creation in CloudWatch. [moby/moby#42132](https://github.com/moby/moby/pull/42132)
- Added a new option to the `awslogs` log driver to specify the log format that is sent to CloudWatch. [moby/moby#42838](https://github.com/moby/moby/pull/42838)
- Added a new option to the `fluentd` log driver to set the reconnection interval. [moby/moby#43100](https://github.com/moby/moby/pull/43100)
- Added new options-setters to the Go API client: `WithTLSClientConfigFromEnv()`, `WithHostFromEnv()`, and `WithVersionFromEnv()`. [moby/moby#42224](https://github.com/moby/moby/pull/42224)

> - Added generation of shell command completion through `docker completion` subcommand. [docker/cli#3429](https://github.com/docker/cli/pull/3429)

Reviewer: ```suggestion
- Added generation of shell command completion through a `docker completion` subcommand. [docker/cli#3429](https://github.com/docker/cli/pull/3429)
```

- API: Added a `Swarm` header to `GET /_ping` and `HEAD /_ping`, allowing single-request detection of Swarm support. [moby/moby#42064](https://github.com/moby/moby/pull/42064)

> - API: Added a `signal` parameter to `POST /containers/{id}/stop` and `POST /containers/{id}/restart` to set the signal used. [moby/moby#43206](https://github.com/moby/moby/pull/43206)

Reviewer 3: We could consider adding back an API changes section, and link to https://docs.docker.com/engine/api/version-history/ "for a full overview of API changes in API v1.43 refer to .."

Author: ACK - I will bring up this suggestion with the docs team

Author: we recently changed the release note template so we're using the same template across all of the products. While it would perhaps be nice to add an API section for the engine section, it doesn't necessarily translate for all other products. So, let's avoid changing the heading now - we can revisit this later. Perhaps we could update the template to make it an optional heading for products where it applies.

- API: Added a `CreateMountPoint` parameter to `POST /containers/create`. [moby/moby#43484](https://github.com/moby/moby/pull/43484)
- API: Added a `shared-size` parameter to `GET /images/json` to enable shared-size computation of images. [moby/moby#42531](https://github.com/moby/moby/pull/42531)
- API: Added a `type` parameter to `GET /system/df`, to control what object types to are considered when computing disk usage. [moby/moby#42559](https://github.com/moby/moby/pull/42559)

> - systemd: Start `docker.service` after `time-set.target`. [moby/moby#43107](https://github.com/moby/moby/pull/43107)

Reviewer: Move this to the `Packaging` section.

Reviewer: We can resolve this as we rejected the Packaging rename.

### Removed

> - Removed support for reading configuration from `~/.dockercfg`. [docker/cli#2504](https://github.com/docker/cli/pull/2504)

Reviewer 3: Similar to the API changes, we could provide a link to the https://docs.docker.com/engine/deprecated/ page for a full overview of deprecated (and/or removed) features in this release. Some thing may be marked deprecated (but not yet removed), and should no longer be used (even if they are currently still functional).

Author: added top-level links to both the API reference and deprecated pages

  This location has been deprecated since 1.7.0.

- Removed the `-g` and `--graph` daemon options in favor of `--data-root`. [docker/cli#3739](https://github.com/docker/cli/pull/3739)

  These options have been hidden and deprecated since 17.05.

- Require explicit configuration to use deprecated storage drivers, and don't automatically select them when upgrading. [moby/moby#43378](https://github.com/moby/moby/pull/43378)
- Removed deprecated support for `overlay` and `overlay2` storage drivers on backing filesystems without `d_type` support. [moby/moby#43472](https://github.com/moby/moby/pull/43472)
- Removed the deprecated `overrideKernelCheck` option from the `overlay2` storage driver. [moby/moby#44279](https://github.com/moby/moby/pull/44279)
- Removed support for the deprecated `io.containerd.runtime.v1.linux` OCI runtime. [moby/moby#43695](https://github.com/moby/moby/pull/43695)

> - Removed LCOW (Linux Containers on Windows). [moby/moby#42451](https://github.com/moby/moby/pull/42451), [moby/moby#42499](https://github.com/moby/moby/pull/42499), [moby/moby#42506](https://github.com/moby/moby/pull/42506), [moby/moby#42511](https://github.com/moby/moby/pull/42511), [moby/moby#42520](https://github.com/moby/moby/pull/42520), [moby/moby#42683](https://github.com/moby/moby/pull/42683), [moby/moby#42684](https://github.com/moby/moby/pull/42684), [moby/moby#42685](https://github.com/moby/moby/pull/42685), [moby/moby#43187](https://github.com/moby/moby/pull/43187)

Reviewer 3: If we link to the deprecated page, we could also consider linking to the relevant section on that page (instead of all the pull requests); in this case, for example, it would be https://docs.docker.com/engine/deprecated/#linux-containers-on-windows-lcow-experimental

Reviewer 3: And for a full list of pull requests / changes included in the release, we can (somewhere) provide links to either;

The milestones;

- https://github.com/docker/cli/milestone/51?closed=1
- https://github.com/moby/moby/milestone/91?closed=1

Or overview of merged PRs filtered by milestone;

- https://github.com/docker/cli/pulls?q=is%3Apr+is%3Amerged+milestone%3A23.0.0
- https://github.com/moby/moby/pulls?q=is%3Apr+is%3Amerged+milestone%3A23.0.0

Author: I started adding links for each item to the relevant section in the deprecated.md, but it doesn't contain all of the entries, so in that case only some of the items in this list will have a link to a deprecated section.

  LCOW was introduced as a technical preview in 17.09 and deprecated in 20.10.

- Removed daemon options related to legacy overlay networks, which were used with "standalone" Swarm.
  - Removed `--cluster-xx` options from `dockerd`. [moby/moby#40383](https://github.com/moby/moby/issues/40383)
  - Removed `host-discovery` and overlay networks with external k/v stores. [moby/moby#42247](https://github.com/moby/moby/pull/42247)
- Removed a deprecated `arm` platform fallback. `--platform linux/arm/vY` will now return a error when `arm/vY` is not available instead of pulling the wrong image. [moby/moby#44414](https://github.com/moby/moby/pull/44414)
- Removed the deprecated `SetCustomHTTPHeaders()`, `CustomHTTPHeaders()` options-setters from the Go client API. [moby/moby#42694](https://github.com/moby/moby/pull/42694)
- Removed the deprecated `WithDialer()` option-setter from the Go client API. [moby/moby#44022](https://github.com/moby/moby/pull/44022)
  
  Users should use `WithDialContext()` instead.

> - Removed the daemon implementation of `opts.QuotedString`, which was moved to CLI. [moby/moby#43250](https://github.com/moby/moby/pull/43250)

Reviewer: ```suggestion
- Removed the daemon implementation of `opts.QuotedString`, which was moved to the CLI. [moby/moby#43250](https://github.com/moby/moby/pull/43250)
```

- Removed separate daemon ID from trust-key in the daemon, and disable generating the trust-key. [moby/moby#43555](https://github.com/moby/moby/pull/43555)
- API: Removed the deprecated `KernelMemory` option from `POST /containers/create` on API >= 1.42. [moby/moby#43214](https://github.com/moby/moby/pull/43214)

### Deprecated

- Require Windows Server RS5 / LTSC 2019 (build 17763) as the minimum to run the daemon. [moby/moby#43254](https://github.com/moby/moby/pull/43254)
- Deprecated `BuilderSize` in API versions >= 1.42 [moby/moby#42608](https://github.com/moby/moby/pull/42608)
- Deprecated `BuildCache.Parent` in favor of the newly introduced `BuildCache.Parents` in API >= v1.42. [moby/moby#43908](https://github.com/moby/moby/pull/43908)
- Deprecated `pkg/urlutil`, moving the implementation to `builder/remotecontext/urlutil`. [moby/moby#43477](https://github.com/moby/moby/pull/43477)

> ### Updates

Reviewer: ```suggestion
### Packaging
```

Reviewer: Seems like the intention might be not renaming this section to stay consistent with other projects, @dvdksn? I think this is still more natural (as we can move the systemd change here), but feel free to resolve this and `s/Packaging/Updates` in my other comments if you think this should stay.

Author: We're trying to keep the release notes aligned across different products, hence my hesitation to change this to Packaging. Though I am not sure I like "Updates" either, so I am changing it to "Upgrades" now.

- Update Go to `1.19.5`. [docker/cli#3958](https://github.com/docker/cli/pull/3958), [moby/moby#44794](https://github.com/moby/moby/pull/44794)
- Update `rootlesskit` to `v0.14.4`. [moby/moby#42708](https://github.com/moby/moby/pull/42708)
- Update `buildkit` to `v0.10.0`. [moby/moby#43239](https://github.com/moby/moby/pull/43239)

> - Update `swarmkit` to `v2.0.0-20221215132206-0da442b2780f`. [moby/moby#44652](https://github.com/moby/moby/pull/44652)

Reviewer: ```suggestion
- Update `swarmkit` to `v2.0.0-20230119195359-904c221ac281`. [moby/moby#44858](https://github.com/moby/moby/pull/44858)
```

- Update `containerd` to `v1.6.15`. [moby/moby#44766](https://github.com/moby/moby/pull/44766), [moby/moby#44769](https://github.com/moby/moby/pull/44769)
- Update `runc` to `v1.1.4`. [moby/moby#44039](https://github.com/moby/moby/pull/44039)
- Update `hcsshim` `v0.9.6`. [moby/moby#44658](https://github.com/moby/moby/pull/44658)

### Security

- Change permissions on container `hostconfig.json` files to `0600` (was `0644`). [moby/moby#41620](https://github.com/moby/moby/pull/41620)

> - Fixed `--seccomp-profile` not accepting `unconfined` and rename the default seccomp profile to `builtin`. [moby/moby#42481](https://github.com/moby/moby/pull/42481)

Reviewer: ```suggestion
- Fixed `--seccomp-profile` not accepting `unconfined` and renamed the default seccomp profile to `builtin`. [moby/moby#42481](https://github.com/moby/moby/pull/42481)
```

- Added support for setting flags passed to `seccomp(2)` in seccomp profiles. [moby/moby#42648](https://github.com/moby/moby/pull/42648)
- Refactored seccomp types to reuse runtime-spec, and add support for `ErrnoRet`. [moby/moby#42005](https://github.com/moby/moby/pull/42005)
- Added support for `DefaultErrnoRet` in `seccomp` profiles. [moby/moby#42604](https://github.com/moby/moby/pull/42604)
- Added an explicit `DefaultErrnoRet` field to the default seccomp profile, with no behavior change. [moby/moby#42649](https://github.com/moby/moby/pull/42649)
- Blocked `socket` with `AF_VSOCK` in the default seccomp profile. [moby/moby#44563](https://github.com/moby/moby/pull/44563)
- Re-enabled `process_vm_readv` and `process_vm_writev` in the default seccomp profile. [moby/moby#42083](https://github.com/moby/moby/pull/42083)
- Added syscalls related to PKU to the default seccomp profile. [moby/moby#43812](https://github.com/moby/moby/pull/43812)
- Allowed `clock_settime64` with `CAP_SYS_TIME`. [moby/moby#43775](https://github.com/moby/moby/pull/43775)
- Allowed `bpf` with `CAP_BPF` and `perf_event_open` with `CAP_PERFMON`. [moby/moby#43988](https://github.com/moby/moby/pull/43988)

> - Explicitly set the `clone3` syscall to return `ENOSYS` in order to ensure `glibc` will correctly fallback to using `clone`. [moby/moby#42681](https://github.com/moby/moby/pull/42681)

Reviewer: ```suggestion
- Explicitly set the `clone3` syscall to return `ENOSYS` in the default seccomp profile, in order to ensure `glibc` will correctly fallback to using `clone`. [moby/moby#42681](https://github.com/moby/moby/pull/42681)
```

### Bug fixes and enhancements

- Promoted `overlay2` to be the default storage driver (`btrfs` and `zfs` are now opt-in). [moby/moby#42661](https://github.com/moby/moby/pull/42661)
- Added spinner loading to `docker cp` command. [docker/cli#2708](https://github.com/docker/cli/pull/2708)
- Deprecated the `ElectAuthServer` function, and made it return the default registry without calling the `GET /info` API endpoint. [docker/cli#2819](https://github.com/docker/cli/pull/2819)
- Progress bars are no longer reversed when rolling back Swarm services [docker/cli#2940](https://github.com/docker/cli/pull/2940)
- Use `net.JoinHostPort()` to fix formatting with IPv6 addresses [docker/cli#2972](https://github.com/docker/cli/pull/2972)

> - CLI: error messages are now printed to `stderr`. [docker/cli#3044](https://github.com/docker/cli/pull/3044)

Reviewer: ```suggestion
- CLI error messages are now printed to `stderr`. [docker/cli#3044](https://github.com/docker/cli/pull/3044)
```

- Removed default value from `--stop-signal` flag, as it may not reflect the actual default used by the daemon. [docker/cli#3245](https://github.com/docker/cli/pull/3245)
- Added Compose schema `3.10` to `docker stack`; allow omitting the `version` field (resulting in `latest`). [docker/cli#3257](https://github.com/docker/cli/pull/3257)
- Made Compose version `3` equivalent to `3.x` (latest) in `docker stack`. [docker/cli#3445](https://github.com/docker/cli/pull/3445)
- Fixed `ctrl-c` hanging on Windows to exit after running a container in non-interactive mode. [docker/cli#3302](https://github.com/docker/cli/pull/3302)
- Added relative source paths to `run` command in `-v`/`--volume` and `-m`/`--mount` flags. [docker/cli#3469](https://github.com/docker/cli/pull/3469)
- Removed the client-side sorting of results, in favor of the order in which the search API returns. [docker/cli#3470](https://github.com/docker/cli/pull/3470)

> - Removed `printStorageDriverWarnings()` function in `cli/command/system`, and documentation updates over deprecation status of `aufs`, `overlay`, `devicemapper`, and `--kernel-memory(-tcp)` [docker/cli#3542](https://github.com/docker/cli/pull/3542)

Reviewer: ```suggestion
- Removed `printStorageDriverWarnings()` function in `cli/command/system`, and update documentation for the deprecation status of `aufs`, `overlay`, `devicemapper`, and `--kernel-memory(-tcp)` [docker/cli#3542](https://github.com/docker/cli/pull/3542)
```

- Quota is no longer disabled on cleanup of the `btrfs` driver. [moby/moby#42273](https://github.com/moby/moby/pull/42273)
- (Accessible) host devices can now be mounted in `--privileged` rootless containers. [moby/moby#42638](https://github.com/moby/moby/pull/42638)
- Fixed incorrect handling of `**/foo` pattern in `fileutils`. [moby/moby#42676](https://github.com/moby/moby/pull/42676)
- Added the ability to pass a `--platform` flag on `docker import` when importing an archive [moby/moby#43103](https://github.com/moby/moby/pull/43103)
- Validation of CPU Realtime options is now performed when the daemon starts instead of performing validations for each individual container, so that we can fail early. [moby/moby#43131](https://github.com/moby/moby/pull/43131)
- Freeze the `namesgenerator` package against new additions. Users will have to be satisfied with the existing 25359 adjective-name combinations. [moby/moby#43210](https://github.com/moby/moby/pull/43210)
- Attach `containers/{ID}/attach/ws` only to streams according by stdin, stdout and stderr parameters. [moby/moby#43322](https://github.com/moby/moby/pull/43322)

> - Fixed UDP traffic in containers not working after the container is restarted on sustained traffic. [moby/moby#43409](https://github.com/moby/moby/pull/43409)

Reviewer: ```suggestion
- Fixed UDP traffic in containers not working after the container is restarted under sustained traffic. [moby/moby#43409](https://github.com/moby/moby/pull/43409)
```

- Added support for pulling images with custom amd64 variants that are used by the latest Go, GCC, LLVM etc. [moby/moby#43434](https://github.com/moby/moby/pull/43434)

> - Improved validation of invalid JSON requests in the API.
>   [moby/moby#43463](https://github.com/moby/moby/pull/43463)

Reviewer: ```suggestion
- Improved validation of invalid JSON requests in the API. [moby/moby#43463](https://github.com/moby/moby/pull/43463)
```
